src/dos/drive_local.cpp

```diff
@@ -560,6 +566,36 @@ bool localFile::Close() {
 		fhandle = 0;
 		open = false;
 	};
+
+	if (newtime) {
+		// backport from DOS_PackDate() and DOS_PackTime()
+		struct tm tim = { 0 };
+		tim.tm_sec  = (time&0x1f)*2;
+		tim.tm_min  = (time>>5)&0x3f;
+		tim.tm_hour = (time>>11)&0x1f;
+		tim.tm_mday = date&0x1f;
+		tim.tm_mon  = ((date>>5)&0x0f)-1;
+		tim.tm_year = (date>>9)+1980-1900;
```

Comment:
Note to myself: verify

Reply:
``` C++
#include <cstdint>
#include <cstdio>
#include <ctime>
#include <list>

using namespace std;

/*
MS-DOS date and MS-DOS time are data formats associated with MS-DOS.
They are used in some file formats from the MS-DOS era. Each is a 
16-bit integer.

MS-DOS date represents a day in the range 1980 to 2099 (or maybe 2107,
but dates after 2099 aren't always correctly handled).

MS-DOS time represents a 2-second interval within some day. The time
is usually expected to be in "local time", and there is no indication
of the time zone. This makes it fairly useless in the internet age.

http://fileformats.archiveteam.org/wiki/MS-DOS_date/time
*/

tm dos_to_local_datetime(const uint16_t sec_count, const uint16_t day_count) {
	tm tim = { 0 };
	tim.tm_sec  = (sec_count & 0x1f) * 2;
	tim.tm_min  = (sec_count >> 5) & 0x3f;
	tim.tm_hour = (sec_count >> 11) & 0x1f;
	tim.tm_mday = day_count & 0x1f;
	tim.tm_mon  = ((day_count >> 5) & 0x0f) - 1;
	tim.tm_year = (day_count >> 9) + 1980 - 1900;
	tim.tm_isdst = -1;
	// serialize time
	mktime(&tim);
	return tim;
}

int main()
{
	std::list<uint16_t> seconds = { 0, 1, 29, 30, 31, 21599, 21600, 21601, 43199, 43200, 43201, 65535 }; 
	std::list<uint16_t> days = { 0, 1, 365, 3653, 7305, 14694, 43465, 46386, 65535 }; 
	for (const auto second : seconds) {
		for (const auto day : days) {
			const tm tim = dos_to_local_datetime(second, day);
			printf("dos_to_local_datetime(%6d, %6d) = %s", second, day, asctime(&tim));
		}
	}
	return 0;
}
```

The above "MSDOS" seconds should represent, after conversion:
- 0,  midnight
- 1,  one second after midnight
- 29, 58 seconds after midnight
- 30, one minute after midnight
- 31, one minute, 2 seconds after midnight
- 21599, 2 seconds before noon
- 21600, noon
- 21601, 2 seconds after noon
- 43199, 2 seconds before midnight
- 43200, midnight
- 43201, 2 seconds into the next day / or equal to 1 above / or undefined?
- 65535, 12 hours, 24 minutes, and 30 seconds into the next day / or rolled over to 12:24:30 / or undefined?

The above days should represent, after conversion:
- 0, Tuesday Jan 1 1980(https://www.google.com/search?q=0+days+after+jan+1+1980)
- 1, Wednesday Jan 2 1980](https://www.google.com/search?q=1+day+after+jan+1+1980)
- 365, Wednesday Dec 31](https://www.google.com/search?q=365+days+after+jan+1+1980)
- 3653, 10 years later, Jan 1 1990](https://www.google.com/search?q=3653+days+after+jan+1+1980)
- 7305, 20 years laters, Jan 1 2000](https://www.google.com/search?q=7305+days+after+jan+1+1980)
- 14694, Wednesday, March 25 2020](https://www.google.com/search?q=14694+days+after+jan+1+1980)
- 43465, Thursday Jan 1 2099](https://www.google.com/search?q=43465+days+after+jan+1+1980)
- 46386, Saturday Jan 1 2107](https://www.google.com/search?q=46386+days+after+jan+1+1980)
- 65535, [Wednesday June 6, 2159](https://www.google.com/search?q=65535+days+after+jan+1+1980)

Output:
```
dos_to_local_datetime(     0,      0) = Fri Nov 30 00:00:00 1979                                                                                                         
dos_to_local_datetime(     0,      1) = Sat Dec  1 00:00:00 1979                                                                                                         
dos_to_local_datetime(     0,    365) = Thu Nov 13 00:00:00 1980                                                                                                         
dos_to_local_datetime(     0,   3653) = Thu Feb  5 00:00:00 1987                                                                                                         
dos_to_local_datetime(     0,   7305) = Sat Apr  9 00:00:00 1994                                                                                                         
dos_to_local_datetime(     0,  14694) = Thu Nov  6 00:00:00 2008                                                                                                         
dos_to_local_datetime(     0,  43465) = Mon Feb  9 00:00:00 2065                                                                                                         
dos_to_local_datetime(     0,  46386) = Thu Sep 18 00:00:00 2070                                                                                                         
dos_to_local_datetime(     0,  65535) = Sat Mar 31 00:00:00 2108                                                                                                         
dos_to_local_datetime(     1,      0) = Fri Nov 30 00:00:02 1979                                                                                                         
dos_to_local_datetime(     1,      1) = Sat Dec  1 00:00:02 1979                                                                                                         
dos_to_local_datetime(     1,    365) = Thu Nov 13 00:00:02 1980                                                                                                         
dos_to_local_datetime(     1,   3653) = Thu Feb  5 00:00:02 1987                                                                                                         
dos_to_local_datetime(     1,   7305) = Sat Apr  9 00:00:02 1994                                                                                                         
dos_to_local_datetime(     1,  14694) = Thu Nov  6 00:00:02 2008                                                                                                         
dos_to_local_datetime(     1,  43465) = Mon Feb  9 00:00:02 2065                                                                                                         
dos_to_local_datetime(     1,  46386) = Thu Sep 18 00:00:02 2070                                                                                                         
dos_to_local_datetime(     1,  65535) = Sat Mar 31 00:00:02 2108                                                                                                         
dos_to_local_datetime(    29,      0) = Fri Nov 30 00:00:58 1979 
￼dos_to_local_datetime(    29,      1) = Sat Dec  1 00:00:58 1979                                                                                                         
dos_to_local_datetime(    29,    365) = Thu Nov 13 00:00:58 1980                                                                                                         
dos_to_local_datetime(    29,   3653) = Thu Feb  5 00:00:58 1987                                                                                                         
dos_to_local_datetime(    29,   7305) = Sat Apr  9 00:00:58 1994                                                                                                         
dos_to_local_datetime(    29,  14694) = Thu Nov  6 00:00:58 2008                                                                                                         
dos_to_local_datetime(    29,  43465) = Mon Feb  9 00:00:58 2065                                                                                                         
dos_to_local_datetime(    29,  46386) = Thu Sep 18 00:00:58 2070                                                                                                         
dos_to_local_datetime(    29,  65535) = Sat Mar 31 00:00:58 2108                                                                                                         
dos_to_local_datetime(    30,      0) = Fri Nov 30 00:01:00 1979                                                                                                         
dos_to_local_datetime(    30,      1) = Sat Dec  1 00:01:00 1979                                                                                                         
dos_to_local_datetime(    30,    365) = Thu Nov 13 00:01:00 1980                                                                                                         
dos_to_local_datetime(    30,   3653) = Thu Feb  5 00:01:00 1987                                                                                                         
dos_to_local_datetime(    30,   7305) = Sat Apr  9 00:01:00 1994                                                                                                         
dos_to_local_datetime(    30,  14694) = Thu Nov  6 00:01:00 2008                                                                                                         
dos_to_local_datetime(    30,  43465) = Mon Feb  9 00:01:00 2065                                                                                                         
dos_to_local_datetime(    30,  46386) = Thu Sep 18 00:01:00 2070                                                                                                         
dos_to_local_datetime(    30,  65535) = Sat Mar 31 00:01:00 2108                                                                                                         
dos_to_local_datetime(    31,      0) = Fri Nov 30 00:01:02 1979                                                                                                         
dos_to_local_datetime(    31,      1) = Sat Dec  1 00:01:02 1979                                                                                                         
dos_to_local_datetime(    31,    365) = Thu Nov 13 00:01:02 1980                                                                                                         
dos_to_local_datetime(    31,   3653) = Thu Feb  5 00:01:02 1987                                                                                                         
dos_to_local_datetime(    31,   7305) = Sat Apr  9 00:01:02 1994                                                                                                         
dos_to_local_datetime(    31,  14694) = Thu Nov  6 00:01:02 2008                                                                                                         
dos_to_local_datetime(    31,  43465) = Mon Feb  9 00:01:02 2065                                                                                                         
dos_to_local_datetime(    31,  46386) = Thu Sep 18 00:01:02 2070                                                                                                         
dos_to_local_datetime(    31,  65535) = Sat Mar 31 00:01:02 2108                                                                                                         
dos_to_local_datetime( 21599,      0) = Fri Nov 30 10:35:02 1979                                                                                                         
dos_to_local_datetime( 21599,      1) = Sat Dec  1 10:35:02 1979                                                                                                         
dos_to_local_datetime( 21599,    365) = Thu Nov 13 10:35:02 1980
dos_to_local_datetime( 21599,   3653) = Thu Feb  5 10:35:02 1987                                                                                                         
dos_to_local_datetime( 21599,   7305) = Sat Apr  9 10:35:02 1994                                                                                                         
dos_to_local_datetime( 21599,  14694) = Thu Nov  6 10:35:02 2008                                                                                                         
dos_to_local_datetime( 21599,  43465) = Mon Feb  9 10:35:02 2065                                                                                                         
dos_to_local_datetime( 21599,  46386) = Thu Sep 18 10:35:02 2070                                                                                                         
dos_to_local_datetime( 21599,  65535) = Sat Mar 31 10:35:02 2108                                                                                                         
dos_to_local_datetime( 21600,      0) = Fri Nov 30 10:35:00 1979                                                                                                         
dos_to_local_datetime( 21600,      1) = Sat Dec  1 10:35:00 1979                                                                                                         
dos_to_local_datetime( 21600,    365) = Thu Nov 13 10:35:00 1980                                                                                                         
dos_to_local_datetime( 21600,   3653) = Thu Feb  5 10:35:00 1987                                                                                                         
dos_to_local_datetime( 21600,   7305) = Sat Apr  9 10:35:00 1994                                                                                                         
dos_to_local_datetime( 21600,  14694) = Thu Nov  6 10:35:00 2008                                                                                                         
dos_to_local_datetime( 21600,  43465) = Mon Feb  9 10:35:00 2065                                                                                                         
dos_to_local_datetime( 21600,  46386) = Thu Sep 18 10:35:00 2070                                                                                                         
dos_to_local_datetime( 21600,  65535) = Sat Mar 31 10:35:00 2108                                                                                                         
dos_to_local_datetime( 21601,      0) = Fri Nov 30 10:35:02 1979                                                                                                         
dos_to_local_datetime( 21601,      1) = Sat Dec  1 10:35:02 1979                                                                                                         
dos_to_local_datetime( 21601,    365) = Thu Nov 13 10:35:02 1980                                                                                                         
dos_to_local_datetime( 21601,   3653) = Thu Feb  5 10:35:02 1987                                                                                                         
dos_to_local_datetime( 21601,   7305) = Sat Apr  9 10:35:02 1994                                                                                                         
dos_to_local_datetime( 21601,  14694) = Thu Nov  6 10:35:02 2008                                                                                                         
dos_to_local_datetime( 21601,  43465) = Mon Feb  9 10:35:02 2065                                                                                                         
dos_to_local_datetime( 21601,  46386) = Thu Sep 18 10:35:02 2070                                                                                                         
dos_to_local_datetime( 21601,  65535) = Sat Mar 31 10:35:02 2108                                                                                                         
dos_to_local_datetime( 43199,      0) = Fri Nov 30 21:06:02 1979                                                                                                         
dos_to_local_datetime( 43199,      1) = Sat Dec  1 21:06:02 1979                                                                                                         
dos_to_local_datetime( 43199,    365) = Thu Nov 13 21:06:02 1980                                                                                                         
dos_to_local_datetime( 43199,   3653) = Thu Feb  5 21:06:02 1987                                                                                                         
dos_to_local_datetime( 43199,   7305) = Sat Apr  9 21:06:02 1994                                                                                                         
dos_to_local_datetime( 43199,  14694) = Thu Nov  6 21:06:02 2008                                                                                                         
dos_to_local_datetime( 43199,  43465) = Mon Feb  9 21:06:02 2065 
dos_to_local_datetime( 43199,  46386) = Thu Sep 18 21:06:02 2070   
dos_to_local_datetime( 43199,  65535) = Sat Mar 31 21:06:02 2108                                                                                                         
dos_to_local_datetime( 43200,      0) = Fri Nov 30 21:06:00 1979                                                                                                         
dos_to_local_datetime( 43200,      1) = Sat Dec  1 21:06:00 1979                                                                                                         
dos_to_local_datetime( 43200,    365) = Thu Nov 13 21:06:00 1980                                                                                                         
dos_to_local_datetime( 43200,   3653) = Thu Feb  5 21:06:00 1987                                                                                                         
dos_to_local_datetime( 43200,   7305) = Sat Apr  9 21:06:00 1994                                                                                                         
dos_to_local_datetime( 43200,  14694) = Thu Nov  6 21:06:00 2008                                                                                                         
dos_to_local_datetime( 43200,  43465) = Mon Feb  9 21:06:00 2065                                                                                                         
dos_to_local_datetime( 43200,  46386) = Thu Sep 18 21:06:00 2070                                                                                                         
dos_to_local_datetime( 43200,  65535) = Sat Mar 31 21:06:00 2108                                                                                                         
dos_to_local_datetime( 43201,      0) = Fri Nov 30 21:06:02 1979                                                                                                         
dos_to_local_datetime( 43201,      1) = Sat Dec  1 21:06:02 1979                                                                                                         
dos_to_local_datetime( 43201,    365) = Thu Nov 13 21:06:02 1980                                                                                                         
dos_to_local_datetime( 43201,   3653) = Thu Feb  5 21:06:02 1987                                                                                                         
dos_to_local_datetime( 43201,   7305) = Sat Apr  9 21:06:02 1994                                                                                                         
dos_to_local_datetime( 43201,  14694) = Thu Nov  6 21:06:02 2008                                                                                                         
dos_to_local_datetime( 43201,  43465) = Mon Feb  9 21:06:02 2065                                                                                                         
dos_to_local_datetime( 43201,  46386) = Thu Sep 18 21:06:02 2070                                                                                                         
dos_to_local_datetime( 43201,  65535) = Sat Mar 31 21:06:02 2108                                                                                                         
dos_to_local_datetime( 65535,      0) = Sat Dec  1 08:04:02 1979                                                                                                         
dos_to_local_datetime( 65535,      1) = Sun Dec  2 08:04:02 1979                                                                                                         
dos_to_local_datetime( 65535,    365) = Fri Nov 14 08:04:02 1980                                                                                                         
dos_to_local_datetime( 65535,   3653) = Fri Feb  6 08:04:02 1987                                                                                                         
dos_to_local_datetime( 65535,   7305) = Sun Apr 10 08:04:02 1994                                                                                                         
dos_to_local_datetime( 65535,  14694) = Fri Nov  7 08:04:02 2008                                                                                                         
dos_to_local_datetime( 65535,  43465) = Tue Feb 10 08:04:02 2065                                                                                                         
dos_to_local_datetime( 65535,  46386) = Fri Sep 19 08:04:02 2070                                                                                                         
dos_to_local_datetime( 65535,  65535) = Sun Apr  1 08:04:02 2108
```